pkg/util/log/audit/audit.go

```diff
@@ -0,0 +1,197 @@
+package audit
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+)
+
+const (
+	MetadataCreatedTime = "createdTime"
+	MetadataPayload     = "payload"
+	MetadataLogKind     = "logKind"
+	MetadataSource      = "source"
+
+	SourceAdminPortal = "aro-admin"
+	SourceRP          = "aro-rp"
+
+	EnvKeyAppID               = "envAppID"
+	EnvKeyAppVer              = "envAppVer"
+	EnvKeyCloudDeploymentUnit = "envCloudDeploymentUnit"
+	EnvKeyCloudRole           = "envCloudRole"
+	EnvKeyCloudRoleVer        = "envCloudRoleVer"
+	EnvKeyCorrelationID       = "envCorrelationID"
+	EnvKeyIKey                = "envIKey"
+
+	PayloadKeyCallerIdentities = "payloadCallerIdentities"
+	PayloadKeyCategory         = "payloadCategory"
+	PayloadKeyNCloud           = "payloadNCloud"
+	PayloadKeyOperationName    = "payloadOperationName"
+	PayloadKeyResult           = "payloadResult"
+	PayloadKeyRequestID        = "payloadRequestID"
+	PayloadKeyTargetResources  = "payloadTargetResources"
+
+	ifxAuditCloudVer = 1.0
+	ifxAuditLogKind  = "ifxaudit"
+	ifxAuditName     = "#Ifx.AuditSchema"
+	ifxAuditVersion  = 2.1
+
+	// ifxAuditFlags is a collection of values bit-packed into a 64-bit integer.
+	// These properties describe how the event should be processed by the pipeline
+	// in an implementation-independent way.
+	ifxAuditFlags = 257
+)
+
+var (
+	// epoch is an unique identifier associated with the current session of the
+	// telemetry library running on the platform. It must be stable during a
+	// session, and has no implied ordering across sessions.
+	epoch = uuid.NewV4().String()
+
+	// seqNum is used to track absolute order of uploaded events, per session.
+	// It is reset when the ARO component is restarted. The first log will have
+	// its sequence number set to 1.
+	seqNum      uint64
+	seqNumMutex sync.Mutex
+)
+
+// NewEntry returns a log entry that embeds the provided logger. It has a hook
+// that knows how to hydrate an IFxAudit log payload before logging it.
+func NewEntry(env env.Core, logger *logrus.Logger) *logrus.Entry {
+	logger.AddHook(&payloadHook{
+		payload: &AuditPayload{},
+		env:     env,
+	})
+
+	return logrus.NewEntry(logger)
+}
+
+// payloadHook, when fires, hydrates an IFxAudit log payload using data in a log
+// entry.
+type payloadHook struct {
+	payload *AuditPayload
+	env     env.Core
+}
+
+func (payloadHook) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+func (h *payloadHook) Fire(entry *logrus.Entry) error {
+	payload := *h.payload // shallow copy
+
+	// Part-A
+	payload.EnvVer = ifxAuditVersion
+	payload.EnvName = ifxAuditName
```

Comment:
@ihcsim the reason for the shallow copy was to try to reduce the number of repeated assignments in Fire().  IIRC you wanted to get all the assignments in one place? (which also makes sense).  Currently the shallow copy is pointless.  Worth working out what to do with this in a follow-up.  I think this probably comes down to the difference between memcpy and individual stores - I'm OK with removing the shallow copy and payload field if you want to?

Reply:
ah.. ok. I thought the pointer is deferenced to avoid race condition. I can remove it in my upcoming PR.